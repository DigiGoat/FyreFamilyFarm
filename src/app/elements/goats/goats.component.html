<!--<app-goat-modal _title="Does" [goats]="does" *ngIf="does.length"></app-goat-modal>-->
<div class="container" *ngIf="!err && !noGoats" test-id="goats">
  <h1 class="text-center" test-id="title">{{name}}</h1>
  <h5 class="text-center" test-id="description">Click on a {{name.slice(0, -1)}} Below For More Info</h5>
  <div class="row row-cols-1 row-cols-md-2 row-cols-xl-3 row-cols-xxl-4 align-items-center justify-content-center">
    <div class="col py-3" *ngFor="let goat of (goats?.length ? goats : undefined)" test-id="goat-cards">
      <app-goat-card [goat]="goat"></app-goat-card>
    </div>
    <div class="col py-3" *ngFor="let goat of (goats?.length ? undefined : [1,2,3,4,5,6,7,8])" test-id="placeholder-cards">
      <app-goat-card></app-goat-card>
    </div>
  </div>
</div>
<div *ngIf="err" test-id="error">
  <h3 class="text-center" test-id="error-title">Failed To Load {{name}} With Error: {{err.status}} - {{err.statusText}}</h3>
  <h6 class="text-center" test-id="error-description">Please Check Your Connection And Try Again Later.</h6>
</div>
<div *ngIf="noGoats" test-id="no-goats">
  <h3 class="text-center" test-id="no-goats-title">No {{name}} Found</h3>
  <h6 class="text-center" test-id="no-goats-description">Please Try Again Later.</h6>
</div>
