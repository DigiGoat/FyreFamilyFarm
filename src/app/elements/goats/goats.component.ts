import type { HttpErrorResponse } from '@angular/common/http';
import { Goat } from 'src/app/services/goat/goat.service';

import { Component, Input, OnInit } from '@angular/core';


import type { Observable } from 'rxjs';


@Component({
  selector: 'app-goats',
  templateUrl: './goats.component.html',
  styleUrls: ['./goats.component.scss']
})
export class GoatsComponent implements OnInit {
  public err?: HttpErrorResponse;
  public noGoats = false;
  constructor() { }
  public goats?: Goat[];
  @Input({ required: true, alias: 'goats' }) getter!: Observable<Goat[]>;
  @Input({ required: true }) name!: string;
  ngOnInit() {
    this.getter.subscribe({
        next: goats => {
        if (!goats.length) {
          this.noGoats = true;
        }
        this.goats = goats;
        },
        error: err => this.err = err
    });
  }
}
