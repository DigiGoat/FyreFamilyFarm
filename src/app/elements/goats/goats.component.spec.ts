import { of } from 'rxjs/internal/observable/of';
import { throwError } from 'rxjs/internal/observable/throwError';

import { ComponentFixture, TestBed } from '@angular/core/testing';

import { GoatsComponent } from './goats.component';


const goat = {
  nickname: 'Your Goats Farm Name',
  name: 'YOUR GOATS ADGA NAME',
  description: 'A quick description about your goat',
  normalizeId: 'PD12345',
  dateOfBirth: new Date(Date.now() - (1000 * 3600 * 24)).toString()
};
describe('GoatsComponent', () => {
  let component: GoatsComponent;
  let fixture: ComponentFixture<GoatsComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [GoatsComponent],
      //schemas: [NO_ERRORS_SCHEMA]
    })
      .compileComponents();
    fixture = TestBed.createComponent(GoatsComponent);
    component = fixture.componentInstance;
    component.name = 'TEST GOATS';
  });
  describe('Without a Goat', () => {
    let html: HTMLElement;
    beforeEach(() => {
      component.getter = of([]);
      fixture.detectChanges();
      html = fixture.nativeElement;
    });
    it('should create', () => {
      expect(component).toBeTruthy();
    });
    it('should set noGoats to true', () => {
      expect(component.noGoats).toBe(true);
    });
    it('should not display goats', () => {
      const element = html.querySelector('[test-id=goats]');
      expect(element).toBeFalsy();
    });
    it('should not display an error', () => {
      const element = html.querySelector('[test-id=error]');
      expect(element).toBeFalsy();
    });
    it('should display no-goats', () => {
      const element = html.querySelector('[test-id=no-goats]');
      expect(element).toBeTruthy();
    });
    it('should display no-goats-title', () => {
      const element = html.querySelector('[test-id=no-goats-title]');
      expect(element?.innerHTML).toBe('No TEST GOATS Found');
    });
    it('should display no-goats-description', () => {
      const element = html.querySelector('[test-id=no-goats-description]');
      expect(element?.innerHTML).toBe('Please Try Again Later.');
    });
    it('should match the snapshot', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
  describe('With a Error', () => {
    let html: HTMLElement;
    beforeEach(() => {
      component.getter = throwError(() => { return { status: -1, statusText: 'TEST_ERROR' }; });
      fixture.detectChanges();
      html = fixture.nativeElement;
    });
    it('should create', () => {
      expect(component).toBeTruthy();
    });
    it('should set err', () => {
      expect(component.err).toEqual({ status: -1, statusText: 'TEST_ERROR' });
    });
    it('should not display goats', () => {
      const element = html.querySelector('[test-id=goats]');
      expect(element).toBeFalsy();
    });
    it('should not display no-goats', () => {
      const element = html.querySelector('[test-id=no-goats]');
      expect(element).toBeFalsy();
    });
    it('should display error', () => {
      const element = html.querySelector('[test-id=error]');
      expect(element).toBeTruthy();
    });
    it('should display no-goats-title', () => {
      const element = html.querySelector('[test-id=error-title]');
      expect(element?.innerHTML).toBe('Failed To Load TEST GOATS With Error: -1 - TEST_ERROR');
    });
    it('should display no-goats-description', () => {
      const element = html.querySelector('[test-id=error-description]');
      expect(element?.innerHTML).toBe('Please Check Your Connection And Try Again Later.');
    });
    it('should match the snapshot', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('With a Goat (v1.0.0)', () => {
    let html: HTMLElement;
    beforeEach(() => {
      component.getter = of([goat]);
      fixture.detectChanges();
      html = fixture.nativeElement;
    });
    it('should create', () => {
      expect(component).toBeTruthy();
    });
    it('should not display no-goats', () => {
      const element = html.querySelector('[test-id=no-goats]');
      expect(element).toBeFalsy();
    });
    it('should not display an error', () => {
      const element = html.querySelector('[test-id=error]');
      expect(element).toBeFalsy();
    });
    it('should display goats', () => {
      const element = html.querySelector('[test-id=goats]');
      expect(element).toBeTruthy();
    });
    it('should display a title', () => {
      const element = html.querySelector('[test-id=title]');
      expect(element?.innerHTML).toBe('TEST GOATS');
    });
    it('should display a description', () => {
      const element = html.querySelector('[test-id=description]');
      expect(element?.innerHTML).toBe('Click on a TEST GOAT Below For More Info');
    });
    it('should display a goat', () => {
      const elements = html.querySelectorAll('[test-id=goat-cards]');
      expect(elements).toBeTruthy();
      expect(elements).toHaveLength(1);
    });
    it('should not display placeholders', () => {
      const elements = html.querySelector('[test-id=goat-placeholders]');
      expect(elements).toBeFalsy();
    });
    it('should match the snapshot', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
  describe('While loading', () => {
    let html: HTMLElement;
    beforeEach(() => {
      component.getter = of([goat]);
      fixture.detectChanges();
      component.goats = [];
      fixture.detectChanges();
      html = fixture.nativeElement;
    });
    it('should create', () => {
      expect(component).toBeTruthy();
    });
    it('should not display no-goats', () => {
      const element = html.querySelector('[test-id=no-goats]');
      expect(element).toBeFalsy();
    });
    it('should not display an error', () => {
      const element = html.querySelector('[test-id=error]');
      expect(element).toBeFalsy();
    });
    it('should display goats', () => {
      const element = html.querySelector('[test-id=goats]');
      expect(element).toBeTruthy();
    });
    it('should display a title', () => {
      const element = html.querySelector('[test-id=title]');
      expect(element?.innerHTML).toBe('TEST GOATS');
    });
    it('should display a description', () => {
      const element = html.querySelector('[test-id=description]');
      expect(element?.innerHTML).toBe('Click on a TEST GOAT Below For More Info');
    });
    it('should display placeholders', () => {
      const elements = html.querySelectorAll('[test-id=placeholder-cards]');
      expect(elements).toBeTruthy();
      expect(elements).toHaveLength(8);
    });
    it('should not display goats', () => {
      const elements = html.querySelector('[test-id=goat-cards]');
      expect(elements).toBeFalsy();
    });
    it('should match the snapshot', () => {
      fixture.detectChanges();
      expect(fixture).toMatchSnapshot();
    });
  });
});
