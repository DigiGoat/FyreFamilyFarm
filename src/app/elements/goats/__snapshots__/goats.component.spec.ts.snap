// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoatsComponent While loading should match the snapshot 1`] = `
<app-goats
  getter={[Function Observable]}
  goats={[Function Array]}
  name={[Function String]}
  noGoats="false"
>
  <div
    class="container"
    test-id="goats"
  >
    <h1
      class="text-center"
      test-id="title"
    >
      TEST GOATS
    </h1>
    <h5
      class="text-center"
      test-id="description"
    >
      Click on a TEST GOAT Below For More Info
    </h5>
    <div
      class="row row-cols-1 row-cols-md-2 row-cols-xl-3 row-cols-xxl-4 align-items-center justify-content-center"
    >
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
      <div
        class="col py-3"
        test-id="placeholder-cards"
      >
        <app-goat-card />
      </div>
    </div>
  </div>
</app-goats>
`;

exports[`GoatsComponent With a Error should match the snapshot 1`] = `
<app-goats
  err={[Function Object]}
  getter={[Function Observable]}
  name={[Function String]}
  noGoats="false"
>
  <div
    test-id="error"
  >
    <h3
      class="text-center"
      test-id="error-title"
    >
      Failed To Load TEST GOATS With Error: -1 - TEST_ERROR
    </h3>
    <h6
      class="text-center"
      test-id="error-description"
    >
      Please Check Your Connection And Try Again Later.
    </h6>
  </div>
</app-goats>
`;

exports[`GoatsComponent With a Goat (v1.0.0) should match the snapshot 1`] = `
<app-goats
  getter={[Function Observable]}
  goats={[Function Array]}
  name={[Function String]}
  noGoats="false"
>
  <div
    class="container"
    test-id="goats"
  >
    <h1
      class="text-center"
      test-id="title"
    >
      TEST GOATS
    </h1>
    <h5
      class="text-center"
      test-id="description"
    >
      Click on a TEST GOAT Below For More Info
    </h5>
    <div
      class="row row-cols-1 row-cols-md-2 row-cols-xl-3 row-cols-xxl-4 align-items-center justify-content-center"
    >
      <div
        class="col py-3"
        test-id="goat-cards"
      >
        <app-goat-card />
      </div>
    </div>
  </div>
</app-goats>
`;

exports[`GoatsComponent Without a Goat should match the snapshot 1`] = `
<app-goats
  getter={[Function Observable]}
  goats={[Function Array]}
  name={[Function String]}
  noGoats={[Function Boolean]}
>
  <div
    test-id="no-goats"
  >
    <h3
      class="text-center"
      test-id="no-goats-title"
    >
      No TEST GOATS Found
    </h3>
    <h6
      class="text-center"
      test-id="no-goats-description"
    >
      Please Try Again Later.
    </h6>
  </div>
</app-goats>
`;
